### ISSUE TITLE
[BOA-05M] Incorrect Validation Logic

### ISSUE THREAD

**Issue by Haypierre** [openfort-xyz/openfort-zksync-contracts#10]
## BOA-05M: Incorrect Validation Logic

| Type | Severity | Location |
| ---- | -------- | -------- |
| [Logical Fault](/appendix/finding-types#logical-fault) | <img class="o-severity o-major" src="https://omniscia.io/report-assets/major.png" /> | [BaseOpenfortAccount.sol:L163-L169](https://github.com/openfort-xyz/openfort-zksync-contracts/blob/e2f09b2c3aaa84686c983b9d42ef3037c2d64dc3/contracts/core/base/BaseOpenfortAccount.sol#L163-L169) |

### Description:

The current [`BaseOpenfortAccount::_validateSignature`](https://github.com/openfort-xyz/openfort-zksync-contracts/blob/e2f09b2c3aaa84686c983b9d42ef3037c2d64dc3/contracts/core/base/BaseOpenfortAccount.sol#L156-L170) validation logic will consider all calls with invalid ECDSA payloads that produce a signer address of `0` to be correct.

### Impact:

All invalid signatures will be considered correct by the current [`BaseOpenfortAccount::_validateSignature`](https://github.com/openfort-xyz/openfort-zksync-contracts/blob/e2f09b2c3aaa84686c983b9d42ef3037c2d64dc3/contracts/core/base/BaseOpenfortAccount.sol#L156-L170) mechanism.

### Example:

```solidity
function _validateSignature(bytes32 _hash, bytes memory _signature, uint256 _to, bytes calldata _data)
    internal
    returns (bytes4 magic)
{
    magic = EIP1271_SUCCESS_RETURN_VALUE;
    address signerAddress = _recoverECDSAsignature(_hash, _signature);

    // Note, that we should abstain from using the require here in order to allow for fee estimation to work
    if (signerAddress != owner() && signerAddress != address(0)) {
        // if not owner, try session key validation
        if (!isValidSessionKey(signerAddress, _to, _data)) {
            magic = "";
        }
    }
}
```

### Recommendation:

We advise the code to assign an `EIP1271_SUCCESS_RETURN_VALUE` to the `magic` variable **solely when verification has been successful** (i.e. `if (signerAddress == owner() || isValidSessionKey(signerAddress, _to, _data)) magic = EIP1271_SUCCESS_RETURN_VALUE;`).

**Comment by Haypierre** [comment]
Remark: 

Description and Impact are incompatible:  only invalid signatures with a hash that recovers to `address(0)` will be considered valid, NOT All invalid signatures.

This has been addressed in #11

**Comment by Haypierre** [comment]
from auditor:


"The ecrecover precompile will yield the zero address on failure which can be trivially crafted:
- https://docs.soliditylang.org/en/v0.8.9/units-and-global-variables.html?highlight=ecrecover#mathematical-and-cryptographic-functions

This is the reason that the standard OpenZeppelin library checks for this scenario here:
- https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/utils/cryptography/ECDSA.sol#L148-L151
"
